README

     C3 version 5:   Cluster Command & Control Suite
         Oak Ridge National Laboratory, Oak Ridge, TN,
    Authors: M.Brim, R.Flanery, G.A.Geist, B.Luethke, S.L.Scott
          Thomas Naughton, Geoffroy Vallee, Wesley Bland

          Copyright (c) 2000-2013    Oak Ridge National Laboratory
                                     All rights reserved.



The Cluster Command and Control (C3) tool suite offers a command line
interface for system and user administration tasks on a cluster. 

 * For help or reporting a bug please e-mail <srt-contact@ornl.gov>.

 * For installation and configuration instructions see INSTALL.

 * For more detailed C3 documentation, see the 'doc/' directory.

 * For examples of scripts using the C3 tools and it's related libraries 
   see the 'contrib/' directory.


==


Note,  node position in the C3 configuration file ('/etc/c3.conf') is very
important!  The numbers used to refer to nodes correspond to their index in
the C3 configuration file (c3.conf), numbering is zero indexed.  C3 version
3.0 and above allows the use of node ranges on the command line.  

To use an example, take the following c3.conf file:

   cluster local {
      htorc-00:node0  #head node
      node[1-64]      #compute nodes
      exclude 60
      node[128-256]
   }

This cluster is made up of 192 nodes. The range of name "node[1-64]" is
listed as slots 0-63 in the list. the range of names "node[128-256] fill up
the slots 64-192. 

As you can see the slot a node fills in a command line range does not
necessarily coincide with it's node name. It is also a reason to explicitly
specify that a node is dead (as opposed to commenting that line out), that
way the node at slot 63 is always the machine labeled node64. 

There are two tools added to the C3 tools suite to help manage node numbers.
cname is given a name and returns its position, or slot, that it occupies.
cnum takes a range and returns what machine names are in that slot. Please
see the c3.conf(5) and c3-range(5) man pages for more details.

Version 4.0 and above allow the use of a scalable configuration. You divide
your cluster into smaller sub-clusters than execute in a parallel. For
example

   cluster part1 {
      node1
      node[1-10]
   }
   cluster part2 {
      node11
      node[11-20]
   }

is a 20 node cluster. Since C3 can easily handle up to about a 64 node
fanout on most hardware the maximum recommended size is 64 64-way clusters,
or 4096 nodes.  See c3-scale(5) for full details.




ChangeLog

	C3 version 5:   Cluster Command & Control Suite
	  Oak Ridge National Laboratory, Oak Ridge, TN,
	Authors: M.Brim, R.Flanery, G.A.Geist, B.Luethke, S.L.Scott
	  Thomas Naughton, Geoffroy Vallee, Wesley Bland

	  Copyright (c) 2000-2013    Oak Ridge National Laboratory
				     All rights reserved.


Version 2.7

	Changed tool names from cl_*(t) to c*(s). 

Version 2.7.1

	Added C3 man page (man c3). 
	Fixed file open error message bug in cpush(s) that misreported that 
		the node list could not be opened, when actually it was the 
		list of files to be pushed which could not be opened. 

Version 2.7.2

	C3 now available as an RPM 
	C3 is now installed in /opt to conform with the Linux Standards Base 
		(LSB) Filesystem Hierarchy Standard (FHS) 
	Updated cpushimage to use SystemImager v1.5.0 
	C3_RSH environment variable should now be set to the path to the 
		binary that C3 should use instead of the default ssh. 
		Previously, C3_RSH was set to 1 if the user wanted to use rsh
		instead of ssh. 

Version 3.0

	Total rewrite of codebase
		using python 2.x
		new config file format
		new command line format
		dropped cps
		dropped serial versions of everthing but cexec

Version 3.1
	
	changed "#!/usr/local/bin/python" to use "/usr/bin/env pyhton2" for 
		portability
	fixed bug where temporary filename stays around (see version 3.0's 
		known bug list)
	when using the cpush -l option if you had more than one space between
		the source and destination cpush failed, you can now have
		any amount and type of white space
	when using cpush with the -l option if there is a blank line execution fails,
		the blank line is now skipped
	now only reads the clusters specified on the command line from the 
		configuration file
	addedd --all option to all the tools
	when using the c3_com_obj.get_clusters() it changed global values instead of local
		values, usefull if using the libraries for your own scripts
	cname can now search multiple names in one command
	when using a direct remote cluster and not using ssh for comminication the scripts
		would still use ssh in one call
	added a "contrib" section of unsupported example scripts (see README
		in contrib directory)
	added c3 command line parser and file parser documentation in the contrib section.
	added the --pipe option, should make it easier to pipe cexec commands
	changed the directory structure of the man pages

Version 3.1.1
	
	--pipe did not handle single line output correctly, now fixed.
	cexec now sorts stderr along with stdout
	*^C no longer prints call stack to screen

Version 3.1.2
	
	fixed cexec where output from a cexec command that is over the system buffer length would hang
	when using cpush with the --head option other command line options were not 
		executed correctly
	fixed a problem where --blind didn't apply when using a file list
	Specifying a cluster on the command line that doesn't exist now gives an appropriate
		error message and tries the next cluster.
	changed the --blind option in cpush to use "rm -f" instead of "rm -rf". This could cause entiere
		directories to be erase (C3 doen not support directoies with cpush)
	when using cnum with an indirect remote cluster it actually executed cname
	added some documentation to the parsing internals for other people to use.
	when using -f or --file with cpush it looked for the alternate configuration file on remote cluster (it
		should have looked for /etc/c3.conf).
	added a --all option to cnum

Version 4.0
	each cluster runs in its own thread, this allows scalable runs of C3 (see README.scalable)
	use of supercluster groups for scalability
	ability to run commands as an alternate user per cluster
	ability to set a different default path to remotly find the c3 commands per user (allows the scalable
		cluster to run entierly in user space)
	ability to specify a different default configuration file per user. (NOT DONE!!!!!!)
	fixed a bug where :torc: would match both the defalut cluster and torc
		instead of being an error
	ability to run as a different user per cluster (see c3-range(5) man page)
	made a few error messages more descriptive
	fixed a bug where if an index on the command line was out of range
		execution quit, it now ignores.

Version 4.1
	Include a new installation mechanism based on a Makefile
	Reduced cpush default verbosity, see C3_VERBOSE EnvVar (from Erich Focht)

Version 5.0
	Internal only release, skipped to v5.0.1 for packaging/distribution reasons.

Version 5.0.1 
	cpush now requires a destination as well as a source (see man pages)
	cpush now supports the use of wildcards on command line (see man pages)
	Changed the way versioning is handled in the utilities and added
	  '--version' option.
	Improved error code handling for the tools.
	Created two new APITests to test return codes for cexec.
	Created new APITests for cpush and updated the old ones to include the 
	   new changes.
	Do not install c3 in /opt/c3-x anymore but directly on the system (e.g.,
	   binaries in /usr/bin).
	Put error codes and default location into a seperate module (include) file
	Added support to crm for deleting multiple files at once.
	Improved internal build system, and added automation of release process.
	Fix for MANPATH (Bernard Li)

Version 5.1.0 
	Fix the man path for Debian packages.
	Include a new binary packages for contrib.
	Update the contrib/c3_conf_manager script so we can add and remove nodes from a given cluster
	Fix bug causing large output to hang cexec - thanks Erik Jacobson (see Ticket#17)
	Require Python >= 2.4, migrate to subprocess module (see Ticket#16)
	Fix ckillnode pid lookup with long process name (see Ticket#15)
	Update copyright stamps to span proper years (2000-2011)

Version 5.1.1 
	Fix a problem in contrib/c3_conf_manager where we were calling delete_cluster instead of del_cluster.

Version 5.1.2 
	Fix a permissions issue on the contrib files.

Version 5.1.3
	Recreate the c3-profiled RPM when running 'make rpm'

src/c3_sock.py
# $Id: c3_sock.py 209 2011-02-02 23:38:27Z tjn $

import os, sys, time
from select import select
from subprocess import *

# this class is used so that I can use strong authentication without
# much hassel and easily implement timeouts. As I am only sending a frame or
# two at a time it is not really noticable. This also allows an abstraction layer 
# so that I can later change to raw socets/PVM/ or whatever without a major code
# re-write: in no way is this meant for speed :)

# one thing to note: these differ from real sockets API in one major way. since the server 
# uses popen to initiate a connextion the clients stdin/stdout is mapped as the socket, thus
# once a client sock has been established all prints will also go to the server, thus you could
# really mess up communications by doing this - be carefull if you use them.

class c3_sock:
	# I use these to abstract whether or not I am writing to stdout/stdin (client)
	# or a pipe (server)
	output_pipe = None
	input_pipe = None
	# default timeout of 10 seconds
	timeout = 10

	# self explanitory
	def set_timeout( self, new_timeout ):
		self.timeout = new_timeout
		
	# general form if a messsage is length:message
	# Simply pack the message and write to a pipe
	def send( self, string_to_send  ):
		length = len( string_to_send )
		string_to_send = str(length) + ':' + string_to_send
		self.output_pipe.write( string_to_send )
		self.output_pipe.flush()


	# these sockets timeout on a receive (think of this as more along the
	# lines of UDP instead of TCP). Thus if you have a client that will sit 
	# a while processing you need some form of stayalive message (I dont
	# do anything that complicated in C3 so I didn't implement them)
	def recieve( self ):
		buffer = ""
		# read a single character in until you get the size terminator
		char_in = self.input_pipe.read( 1 )
		time_start = time.time()
		time_elapsed = 0
		while char_in != ':':
			buffer = buffer + char_in
			char_in = self.input_pipe.read( 1 )
			# as you can see timeouts work but are kinda cheesy, use a nonblocking
			# read and increment a time counter if nothing is retuned, raise and exception
			# if a threshold is crosed.
			time_elapsed = time.time() - time_start
			if char_in != "":
				time_start = time.time()
			if char_in == "" and time_elapsed > self.timeout:
				raise 'time_out'

		# try and read the message from the pipe in as large a chunk as possible, loop
		# untill stated length is reached
		length = int (buffer)
		buffer = self.input_pipe.read( length )
		time_start = time.time()
		time_elapsed = 0
		while len(buffer) < length:
			new_buffer = buffer + self.input_pipe.read( length )
			if len(new_buffer) > len(buffer):
				time_start = time.time()
				time_elapsed = 0
				buffer = new_buffer
			elif time_elapsed > self.timeout:
				raise 'time_out'
			else:
				time_elapsed = time.time() - time_start
		# return message received through pipe		
		return buffer

	def close( self ):
		self.__del__()

# server sock just defines two pipes (input, output) and initializes a command
# this this socket is a socket to a command. Think "ssh node10 client_code"
# where client code communicates through stdin and stdout.
# when it's done simply close pipes
class server_sock( c3_sock ):

	def __init__(self, command):
		#self.output_pipe, self.input_pipe = os.popen2( command )
		# XXX: Changed to subprocess module (see ticket:16)
		p = Popen(command, shell=True, stdin=PIPE, stdout=PIPE, close_fds=True)
		(self.output_pipe, self.input_pipe) = (p.stdin, p.stdout)



	def __del__( self ):
		self.output_pipe.close()
		self.input_pipe.close()
		
# the client sock takes stdin/stdout and map them to another pipe, this is so
# the same send/receive function can be used. Since the socket uses only
# stdin/stdout to communicate only one socket can be running at a time per
# instance of the PROGRAM. stdout and stdin are set to None to catch this error.
# on socket closing stdin and stdout are set back to thier original values.
class client_sock( c3_sock ):

	def __init__(self):
		if sys.stdout == None and sys.stdin == None:
			raise 'single client only'
		self.output_pipe = sys.stdout
		self.input_pipe = sys.stdin
		sys.stdout = None
		sys.stdin = None



	def __del__( self ):
		sys.stdin = self.input_pipe
		sys.stdout = self.output_pipe

# vim:tabstop=4:shiftwidth=4:noexpandtab:textwidth=76


src/c3_except.py
# $Id: c3_except.py 186 2011-01-21 23:07:00Z tjn $

class c3_except( Exception ):
	description = ""
	last = ""

	def __init__(self, string, name):
		self.description = string
		self.last = name


class parse_error( c3_except ):
	pass

class bad_cluster_name( parse_error ):
	pass

class no_more_clusters( c3_except ):
	pass

class invalid_head_node( parse_error ):
	pass

class invalid_cluster_block( parse_error ):
	pass

class no_head_node( c3_except ):
	pass

class invalid_node( parse_error ):
	pass

class end_of_cluster( c3_except ):
	pass

class internel_error( c3_except ):
	pass

class indirect_cluster( parse_error ):
	pass

class not_in_range( parse_error ):
	pass

class end_of_option( c3_except ):
	pass

class bad_string( parse_error ):
	pass

class end_of_opt_string( bad_string ):
	pass

# vim:tabstop=4:shiftwidth=4:noexpandtab:textwidth=76


src/c3_com_obj.py
# $Id: c3_com_obj.py 186 2011-01-21 23:07:00Z tjn $

import re, sys
from c3_except import *

# class used to represent cluster taken from command line
# clusters is the list of cluster names used as a key in the
# nodes associative array. To iterate through them would look as such:
# for cname in c3_cluster_list.clusters:
#	for node in c3_cluster_list[cname]:
#		print node

class c3_cluster_list:
	def __init__(self):
		self.clusters = []
		self.node = {}
		self.username = {}
class c3_command_line:


	# matches any non whitespace word
	any_token = re.compile( r"\s*(?P<word>\S+)" )

	# matches a - or -- and then the option name
	option = re.compile( r"\s*(?P<word>--?\w+)" )
	
	# matches a cluster name
	c_name = re.compile( r"\s+(?P<username>[\w_\-]+@)?(?P<name>[\w_\-]+)?:" );
	
	# matches a single number
	number = re.compile( r"\s*(?P<num>\d+)" );
	
	# matches a range quaifier "-"
	range_qual = re.compile( r"\s*-" );
	
	# matches a single node qualifier ","
	single = re.compile( r"\s*," );
 
	# initiliaze line as an enpty string
	line = ""


	# initilaize internal line as the text to be parsed
	def __init__( self, command_line ):
		self.line = command_line

	# returns a single option, a - or -- followed by string
	def get_opt( self ):
		match = self.option.match( self.line )
		if match:
			line_out = ""
			line_out = match.group( "word" )
			self.line = self.line[match.end():]
		else:
			raise end_of_option( None, None )
		return line_out

	# returns a single token, used for cases such as --file FILENAME to get FILENAME
	def get_opt_string( self ):
		match = self.any_token.match( self.line )
		if match:
			line_out = ""
			line_out = match.group( "word" )
			self.line = self.line[match.end():]
		elif len( self.line ) == 0:
			raise end_of_opt_string( "option needs a string", None)
		else:
			raise bad_string( "option requires a string", None )
		return line_out

	# all parsing is done externally through this command
	def get_clusters( self ):
		cluster_obj = c3_cluster_list()

		# check if any clusters are specified on the command line, if not
		# then set execution to default cluster (name not known at this point
		# /default is an invalid cluster name and hence a place holder in this
		# context
		match = self.c_name.match( self.line )
		if not match:
			cluster_obj.clusters.append( "/default" )
			cluster_obj.node["/default"]=[]
			cluster_obj.node["/default"].append( "" )
			cluster_obj.username["/default"]="/default"
		# while there are still cluster blocks on the command line
		while match:
			# string parsed text from line
			self.line = self.line[match.end():]
			# if a name is specified get it, else : is specified set to default cluster
			if match.group( "name" ):
				index = match.group( "name" )
			else:
				index = "/default"

			#if an alternate username is specified use it.
			if match.group( "username" ):
				cluster_obj.username[index]=match.group( "username" )[:-1]
			else:
				cluster_obj.username[index]="/default"

			# add name to cluster list and initialize node list	
			cluster_obj.clusters.append( index )

			match = self.number.match( self.line )
			cluster_obj.node[index] = []
			cluster_obj.node[index].append( "" )
			node_index = 0
			
			# if a range has been specified on command line parse it
			# this process gets the first number from the list and stores it in a temp
			# var. Then checks if it is part of a range or a single number. If it is a range 
			# it processes once and then checks for a single number, if it is a single number
			# it processes single numbers until wither the end of the node position specification
			# or a range qualifier is found, if a range is found the process loops. In this way the 
			# would be parsed correctly: "1,3,4-8,10-20,25"
			while match:
				# strip parsed text from line
				self.line = self.line[match.end():]
				# get starting number from match
				cluster_obj.node[index][node_index] = int( match.group( "num") )
				# if a match is specified parse it
				match = self.range_qual.match( self.line )
				if match:
					# strip parsed text from line
					self.line = self.line[match.end():]
					match = self.number.match( self.line )
					# set start and end of range
					start_range = cluster_obj.node[index][node_index] + 1
					if match:
						end_range = int( match.group( "num" ) ) + 1
					else:
						self.line = '-' + self.line
						return cluster_obj
					# loop from start to end and add node position to node list
					for counter in  range( start_range, end_range ):
						node_index = node_index + 1
						cluster_obj.node[index].append( "" )
						cluster_obj.node[index][node_index] = counter
					# strip parsed text from line
					self.line = self.line[match.end():]
				# check for single numbers (2,5,6)
				match = self.single.match( self.line )
				if match:
					# srip parsed text from line
					self.line = self.line[match.end():]
					# add node poistion to node list
					cluster_obj.node[index].append( "" )
					node_index = node_index + 1
					match = self.number.match( self.line )
			match = self.c_name.match( self.line )
		try:
			if self.line[0] != ' ':
				raise bad_cluster_name ( None, None )
		except IndexError:
			pass
		return cluster_obj

	# returns the rest of line (usually for the end of the command)
	def rest_of_command( self ):
		return self.line	

# vim:tabstop=4:shiftwidth=4:noexpandtab:textwidth=76


src/c3_file_obj.py
# $Id: c3_file_obj.py 186 2011-01-21 23:07:00Z tjn $

import sys, re
from c3_except import *

# this is the internal representation of a node
# as of right now it is just simply it's name
# (ip or alias) and it's status. dead = 0 is alive
# dead = 1 id offline
class node_obj:
	name = ""
	dead = 0

class cluster_def:
	"parses the cluster definition file"

	###################################################################
	#  this is the regular expressions used parse the file	    #
	###################################################################

	# beginning of cluster definition (name of the cluster)
	# matches cluster_name {
	cluster_name = re.compile ( r"""
		\s*cluster\s+		#cluster keyword
		(?P<c_name>		#cluster name
			[\w_\-]+	#may contain an alphanumeric, underscore, and dash
		)\s*{[\t\v ]*(\#.*)?\n""", re.VERBOSE | re.IGNORECASE
	)

	# extracts the name of the head node
	# matches external:internal 
	# with the internal name optional
	head_node = re.compile ( r"""
		\s*			#get rid of whitespace
		(?P<extname>		#external head node name goes first
			[\w\-\.]+
		)?
		(:			#if internal name spcified
			(?P<intname>	#extract internal name
				[\w\-\.]+
			)
		){0,1}[\t\v" "]*(\#.*)?\n""", re.VERBOSE
	)

	# extracts the name of compute nodes
	# matches dead nodename 
	# with dead being optional and including
	# ranges
	compute_node = re.compile ( r"""
		\s*	#get rid of whitespace
		(?P<dead_node>
			dead[\t\v" "]+
		)?
		(?P<comname>	#get name of current node
			[\w\-.]+		#non range part of name
		)
		(?P<range>
			\[(?P<start>\d+)\-(?P<stop>\d+)\]	#get range (optional)
		){0,1}[\t\v" "]*(\#.*)?\n""", re.VERBOSE
	)


	# exclude nodes from a range
	# matches exclude [num1-num2]
	# with num1 and num2 being integers
	exclude = re.compile ( r"""
		\s*exclude\s*
		((?P<single>\s+\d+)|(\[(?P<start>\d+)\-(?P<stop>\d+)\]))
		[\t\v" "]*(\#.*)?\n
		""", re.VERBOSE | re.IGNORECASE
	)

	# matches brackets { }
	start_bracket = re.compile( """\s*{""" )
	end_bracket = re.compile ( """\s*}[\t\v" "]*""" )

	# matches any non whitespace character
	any_token = re.compile ("\s*\S+")

	# matches a comment line
	comment = re.compile( r"[ \t\r\f\v]*#.*\n" )

	#########################################################
	# variables needed for execution			#
	#########################################################

	# filename of cluster config file
	file = ""

	# this is a string used to hold the config file
	line = ""

	# string to hole the current cluster name
	c_name = ""

	#strings to hold the head node names
	head_int = ""
	head_ext = ""

	#list to hold ranges for nodes
	node_list = []

	#used to show place where error occured
	last_cluster = "first cluster in list"
	last_machine = "first node in list"

	#########################################################
	#							#
	# this is the constructor, it takes the filename of the #
	# config file to parse, the second init throws an error #
	# if no file name is given				#
	#							#
	#########################################################
	def __init__(self, filename):

		self.file = filename
		inputfile = open( filename, "r" )

		# generate a string containing the file 
		line_in = inputfile.readline()
		while line_in:
			self.line = self.line + line_in
			line_in = inputfile.readline()
		inputfile.close()

	#########################################################
	# resets the internal variables after an error		#
	#########################################################
	def reset_vars(self):
		self.line	= 	""
		self.c_name 	= 	""
		self.head_int	=	""
		self.head_ext	=	""
		self.node_list	=	[]
		self.last_cluster = 	"first cluster in list"
		self.last_machine = 	"first node in list"

	#########################################################
	# re-initializes the file (begins at the first cluster  #
	# again)						#
	#########################################################
	def reread_file(self):
		self.reset_vars()
		self.__init__(self.file)

	#########################################################
	# strips comments from front of file			#
	#########################################################
	def strip_comments( self ):

		match = self.comment.match( self.line )
		while match:
			self.line = self.line[match.end():]
			match = self.comment.match( self.line )

	#########################################################
	# scans to next cluster in the file, if called for the  #
	# first time goes to first cluster			#
	# doesn't return anything, just sets internal variable	#
 	#########################################################
	def get_next_cluster(self):

		self.strip_comments()
		match = None

		try:
			while not match:  #loop untill a cluster tag is found
				match = self.cluster_name.match( self.line )
				if match: #if cluster tag found
					# get cluster name
					self.c_name = match.group( "c_name" )
					self.line = self.line[match.end():]
					self.strip_comments()
					self.last_cluster = self.c_name 
					try:
						match = self.head_node.match( self.line )
						if not match.group( "extname" ):
							# this indicates that it is an "indirect" cluster
							# the internal node is actually the external link
							# it was done this way because with normal operation
							# this would be an impossible state 
							self.head_ext = None
							self.head_int = match.group( "intname" )
						else: # "direct" cluster
							self.head_ext = match.group( "extname" )
							self.head_int = match.group( "intname" )
							if not self.head_int:
								self.head_int = self.head_ext
					except AttributeError: # parse error on the head node
						name = self.c_name
						self.reset_vars()
						raise invalid_head_node( "invalid head node specification", name)
					self.line = self.line[match.end():]
					self.strip_comments()
				else: # cluster tag not found
					# strip a single token from self.line
					match = self.any_token.match( self.line );
					# an open bracket here would mean that a valid cluster tag was not found
					# but a new cluster block was trying to be formed
					if self.start_bracket.match( match.group() ):
						name = self.last_cluster
						self.reset_vars()
						raise invalid_cluster_block( "invalid cluster definition", name)
					self.line = self.line[match.end():]
					self.strip_comments()
					match=None
		except AttributeError: # invalid cluster definition
			name = self.last_cluster
			self.reset_vars()
			raise no_more_clusters( "No more valid cluster blocks", name )

	#########################################################
	# returns the external name of the current cluster 	#
	# being parsed						#
	#########################################################
	def get_external_head_node(self):
		if self.head_ext == "":
			raise no_head_node( "no head node set.", "no cluster read yet." )
		return self.head_ext;

	#########################################################
	# returns the internal name of the current cluster 	#
	# being parsed						#
	#########################################################
	def get_internal_head_node(self):
		if self.head_int == "":
			raise no_head_node( "no head node set.", "no cluster read yet." )
		return self.head_int;

	#########################################################
	# returns the name of the next node in the files if 	#
	# called for the first time returns the first node name #
	# returns a node_obj with the appropriate values filed	#
	# in							#
	#########################################################
	def get_next_node(self):
		self.strip_comments()

		# the only time it is possible for this to occur is
		# with indirect clusters
		if self.head_ext == None:
			name = "cluster " + self.c_name
			raise indirect_cluster( "indirect clusters don't have nodes", name )
		node_out = node_obj()

		# when a range is specified a queue is built with the nodes
		# so if a queue is present then we know that a range has
		# been specified and must be used up before we parse another
		# line in the file
		if self.node_list:
			node_out = self.node_list.pop(0)
			self.last_machine = node_out.name
			return node_out

		match = self.compute_node.match( self.line )
 
		if match: # if a compute node is found

			self.line = self.line[match.end():]
			self.strip_comments()
			if match.group( "dead_node" ): # check if it is a dead node 
				if not match.group( "range" ): # dead node qualifier invalid with a range
					node_out.name = match.group( "comname" )
					node_out.dead = 1
				else: # return the given node with a dead set to true
					name = self.last_machine + " in " + self.c_name
					self.reset_vars()
					raise invalid_node( "dead specifier can not have a range", name )
			else: # either a range or single node specified
				if match.group( "range" ): # if range
					# retrieve starting and stopping ranges
					start_add_range = int( match.group( "start" ) )
					stop_add_range = int( match.group( "stop" ) ) + 1
					# start is always zero - start_add_range - start_add_range
					# this is done so that the indexing starts at zero
					start = 0
					stop = stop_add_range - start_add_range
					for index in range(start, stop): # populate list
						self.node_list.append( node_obj() )
						self.node_list[index].name = match.group( "comname" ) + str( index + start_add_range )
						self.node_list[index].dead = 0
					match = self.exclude.match( self.line )
					# multiple exclude lines after a range are valid, hence the while loop
					while match and self.node_list: 
						try:
							if match.group( "single" ): # excluding a single machine
								index = int( match.group( "single" ) ) 
								if index < 0:
									raise IndexError
								self.node_list[index - start_add_range].dead = 1
							else: # excluding a range
								start_ex_range = int( match.group( "start" ) )
								stop_ex_range = int( match.group( "stop" ) )
								# list index starts at zero so the exclude index and
								# list index must co-incide
								start = start_ex_range - start_add_range
								stop = (stop_ex_range - start_add_range) + 1
								if (start <  0) or (start > stop):
									raise IndexError
								for index in range (start, stop):
									self.node_list[index].dead = 1
							self.line = self.line[match.end():]
							self.strip_comments()
							match = self.exclude.match( self.line) # check for second exclude
						except IndexError:
							name = self.last_machine + " in " + self.c_name
							raise not_in_range( "index in exclude is not in range", name)

					node_out = self.node_list.pop(0)
				else: # single node specifier 
					node_out.name = match.group( "comname" )
					node_out.dead = 0
		else: # either there are no more nodes ( closing bracket is found )
		      # or there was a parse error on the node specification line
			if self.end_bracket.match( self.any_token.match(self.line).group() ):
				raise end_of_cluster( "no more nodes in config file", None )
			name = self.last_machine + " in " + self.c_name
			self.reset_vars()
			raise invalid_node( "invalid specification for a node", name )
		self.last_machine = node_out.name
		return node_out

	#########################################################
	# returns the name of the current cluster being read    #
	#########################################################
	def get_cluster_name(self):
		if self.c_name == "":
			raise no_cluster_name( "read in a cluster before useing", "no cluster read yet" )
		return self.c_name

# vim:tabstop=4:shiftwidth=4:noexpandtab:textwidth=76
